src/tools/simeng/gdb/GDBStub.cc

```diff
@@ -0,0 +1,331 @@
+
+  return output;
+}
+
+std::string GDBStub::generateReply(std::string packet) const {
+  std::string output = "$" + packet + "#" + computeChecksum(packet);
+
+  return output;
+}
+
+std::string GDBStub::handleReadRegisters() const {
+  simeng::ArchitecturalRegisterFileSet registers =
+      core_.getArchitecturalRegisterFileSet();
+  std::string output = "";
+
+  for (uint16_t i = 0; i < 32;
+       i++) {  // general purpose registers 0-31, 31 = stack pointer
+    uint64_t value = registers.get({0, i}).get<uint64_t>();
+    output += decToRSP(value);
+  }
+
+  output += decToRSP(core_.getProgramCounter());  // program counter
+
+  output += byteToHex(
+      registers.get({3, 0}).get<uint8_t>());  // NZCV (first 4 bits of cpsr)
+
+  output += "000000";  // rest of cpsr, unneeded
+
+  return output;
+}
+
+std::string GDBStub::handleReadSingleRegister(std::string registerName) const {
+  uint16_t registerNumber = hexToInt(registerName);
+  uint64_t value = core_.getArchitecturalRegisterFileSet()
+                       .get({0, registerNumber})
+                       .get<uint64_t>();
+
+  return decToRSP(value);
+}
+
+std::string GDBStub::handleReadMemory(std::string hexAddress,
+                                      std::string length) const {
+  int numberOfBytes = std::stoi(length);
+  uint64_t intAddress = hexToInt(hexAddress);
+
+  char* memoryPointer = dataMemory_.getMemoryPointer();
+  const char* ptr = memoryPointer + intAddress;
+  uint8_t dest[numberOfBytes];
+
+  if (verbose_)
+    std::cout << "   Reading " << numberOfBytes << " bytes from memory address "
+              << intAddress << std::endl;
+
+  memcpy(dest, ptr, numberOfBytes);
+
+  std::string output;
+  for (uint8_t byte : dest) output += byteToHex(byte);
+
+  return output;
+}
+
+void GDBStub::handleCreateBreakpoint(std::string type, std::string address) {
+  if (stoi(type) > 1)
+    std::cout << RED << "   Watchpoints not supported, no breakpoint created\n"
+              << RESET;
+
+  breakpoints_.push_back(address);
+  if (verbose_)
+    std::cout << "   Breakpoint created at address 0x" << address << std::endl;
+}
+
+void GDBStub::handleRemoveBreakpoint(std::string type, std::string address) {
+  if (stoi(type) > 1)
+    std::cout << RED << "   Watchpoints not supported, no breakpoint removed\n"
+              << RESET;
+
+  for (uint i = 0; i < breakpoints_.size(); i++) {
+    if (breakpoints_[i] == address)
+      breakpoints_.erase(breakpoints_.begin() + i);
+  }
+  if (verbose_)
+    std::cout << "   Breakpoint(s) removed at address 0x" << address
+              << std::endl;
+}
+
+std::string GDBStub::handleContinue() {
+  if (breakpoints_.size() == 0) return "";
+
+  bool hitBreakpoint = 0;
+
+  while (!hitBreakpoint) {
+    core_.tick();
+    dataMemory_.tick();
+    for (std::string breakpoint : breakpoints_) {
+      if (hexToInt(breakpoint) == core_.getProgramCounter()) hitBreakpoint = 1;
+    }
+  }
+
+  return "S05";
```

Comment:
May be worth defining this string using `#define` as it is used in a few places

Reply:
I'm using it 3 times - what's the benefit of using a `#define` for it, and what would be a useful name for it ?

Reply:
If it needed changing in the future then it ensures that the value is changed in all places it is used. Although that may not be relevant to this exact case where this string probably won't change, it is good practice to do so.

As for a name, whatever "S05" represents within GDB

Reply:
cool - i'll do that.

I'm also using "OK" as a common response to GDB, should i `#define` that too ?

Reply:
Could be useful yes